core/except.cpp

```diff
@@ -292,6 +292,17 @@ void validSlice(const Dimensions &dims, const Slice &slice) {
     throw except::SliceError("Expected " + to_string(slice) + " to be in " +
                              to_string(dims) + ".");
 }
+void validSlice(const std::unordered_map<Dim, scipp::index> &dims,
+                const Slice &slice) {
+  if (dims.find(slice.dim) == dims.end() || slice.begin < 0 ||
+      slice.begin >=
+          std::min(slice.end >= 0 ? slice.end + 1 : dims.at(slice.dim),
+                   dims.at(slice.dim)) ||
+      slice.end > dims.at(slice.dim))
```

Comment:
I suggest to move some of these checks into the constructor of `Slice`, then you get the benefit everywhere (and this can avoid misunderstandings like the one caused by interpreting the `-1` as in Python, as you pointed out rightly):
- begin must be non-negative
- custom end should not be smaller than begin
- `end = -1` should be rejected, creation with this special value should be done with a constructor that does not have `end`,

Reply:
I completely agree. I think making the constructor of Slice richer so we can prevent errors closer to the point they are make will save us a lot of future debugging headaches.